gquant/dataframe_flow/_node_flow.py

```diff
@@ -508,6 +508,9 @@ def flow(self):
                 else:
                     df = output_df
 
+            if isinstance(df, dask_cudf.DataFrame):
+                df = df.persist()
```

Comment:
This `persist` call is causing some strange error in notebook "04_portfolio_trade.ipynb". During the call in that notebook:
```
%%time
o_dask = task_graph.run(
            outputs=['sharpe_ratio', 'cumlative_return', 'load_csv_data', 'sort_2'],
            replace={'load_csv_data': {"type": "DaskCsvStockLoader",
                                       "conf": {"path": "many-small"}},
                     'filter_value': {"conf": [{"column": "volume_mean", "min": min_volume},
                                               {"column": "returns_max", "max": max_rate},
                                                {"column":  "returns_min", "min": min_rate}]}}, profile=True)

dask_input_cached = o_dask[2]  # 'load_csv_data' node output
dask_strategy_cached = o_dask[3]  # 'sort_2' node output
```

I get this error although the cell completes and the cells afterward run as well:
```
  File "/opt/conda/envs/rapids/lib/python3.6/site-packages/distributed/protocol/serialize.py", line 210, in serialize
    raise TypeError(msg, str(x)[:10000])
TypeError: ('Could not serialize object of type function.', '<function NodeTaskGraphMixin.decorate_process.<locals>.timer at 0x7fb958c8aea0>')
distributed.comm.utils - ERROR - ('Could not serialize object of type function.', '<function NodeTaskGraphMixin.decorate_process.<locals>.timer at 0x7fb958c8aea0>')


  File "/opt/conda/envs/rapids/lib/python3.6/site-packages/distributed/protocol/serialize.py", line 210, in serialize
    raise TypeError(msg, str(x)[:10000])
TypeError: ('Could not serialize object of type function.', '<function NodeTaskGraphMixin.decorate_process.<locals>.timer at 0x7fb958c8aea0>')

```

Without persisting everything runs fine as well, but the above error does not occur.

Please take a look. Is that `persist` needed here?

Reply:
It works for me
<img width="1240" alt="image" src="https://user-images.githubusercontent.com/43824965/86297420-49406900-bbc9-11ea-93ae-f01fd1819721.png">
Yes. `persist` is the key to speed up the Dask calculation. It shorten the time from 1-2mins to 17seconds. 
Can you try set the flag `profile=False` to see you still have the problem?

Reply:
The `profile=False` does not seem to be the problem. That particular error `Could not serialize object of type function` I'm getting inconsistently. Sometimes it occurs and sometime it doesn't.

Nevertheless, I believe putting the `df = df.persist()` in this particular location is a bug. I ran the other notebooks and this caused notebooks "05_customize_nodes.ipynb" and "05b_customize_nodes_with_ports.ipynb" to crash. This change causes mutation of the dask dataframe.

I debugged it and figured out the problem. That persist logic needs to be handled carefully. I added another method in class `NodeTaskGraphMixin`.
```
    def __delayed_call_noports(self, inputs):

        def get_pout(df_out):
            '''Used for delayed unpacking.'''
             if isinstance(df_out, cudf.DataFrame):
                # Needed for the same reason as __make_copy. To prevent columns
                # addition in the input data frames. In python everything is
                # by reference value and dataframes are mutable.
                # Handle the case when dask_cudf.DataFrames are source frames
                # which appear as cudf.DataFrame in a dask-delayed function.
                return df_out.copy(deep=False)

            return df_out

        # handle the dask dataframe automatically
        # use the to_delayed interface
        # TODO, currently only handles first input is dask_cudf df
        i_df = inputs[0]
        rest = inputs[1:]
        if isinstance(i_df, dask_cudf.DataFrame):
            output_df_dly_list = []
            for input_dly in i_df.to_delayed():
                inputs_ = [input_dly] + rest
                output_df_dly = dask.delayed(self.decorate_process())(inputs_)
                output_df_dly_per = output_df_dly.persist()
                df_out = dask.delayed(get_pout)(output_df_dly_per)
                output_df_dly_list.append(df_out.persist())

            output_df = dask_cudf.from_delayed(output_df_dly_list)

        else:
            output_df = self.decorate_process()(inputs)

        return output_df
```

And then change the logic in `__call__` from:
```
                else:
                    # handle the dask dataframe automatically
                    # use the to_delayed interface
                    # TODO, currently only handles first input is dask_cudf df
                    i_df = inputs[0]
                    rest = inputs[1:]
                    if isinstance(i_df, dask_cudf.DataFrame):
                        d_fun = dask.delayed(self.decorate_process())
                        output_df = dask_cudf.from_delayed([
                            d_fun([item] + rest)
                            for item in i_df.to_delayed()])
                    else:
                        output_df = self.decorate_process()(inputs)
```
TO:
```
                else:
                    output_df = self.__delayed_call_noports(inputs)
```

This will fix the issue and retain performance in "04_portfolio_trade.ipynb". Using the ports API already handles the `persist` logic, but it was not being handled for non-ports case.

Reply:
Great. It fixes the problem.